utils/zerovec/src/varzerovec/slice.rs

```diff
@@ -60,41 +61,136 @@ pub struct VarZeroSlice<T: ?Sized> {
 }
 
 impl<T: VarULE + ?Sized> VarZeroSlice<T> {
-    /// Obtain a [`VarZeroVecBorrowed`] borrowing from the internal buffer
+    /// Construct a new empty VarZeroSlice
+    pub fn new_empty() -> &'static Self {
+        let arr: &[u8] = &[];
+        unsafe { mem::transmute(arr) }
+    }
+    /// Obtain a [`VarZeroVecComponents`] borrowing from the internal buffer
     #[inline]
-    pub fn as_borrowed<'a>(&'a self) -> VarZeroVecBorrowed<'a, T> {
+    pub(crate) fn as_components<'a>(&'a self) -> VarZeroVecComponents<'a, T> {
         unsafe {
             // safety: VarZeroSlice is guaranteed to parse here
-            VarZeroVecBorrowed::from_bytes_unchecked(&self.entire_slice)
+            VarZeroVecComponents::from_bytes_unchecked(&self.entire_slice)
         }
     }
 
-    /// Get the number of elements in this vector
+    /// Get the number of elements in this slice
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// # use std::str::Utf8Error;
+    /// # use zerovec::ule::ZeroVecError;
+    /// # use zerovec::VarZeroVec;
+    ///
+    /// let strings = vec!["foo", "bar", "baz", "quux"];
+    /// let bytes = VarZeroVec::<str>::get_serializable_bytes(&strings).unwrap();
+    ///
+    /// let mut vec: VarZeroVec<str> = VarZeroVec::parse_byte_slice(&bytes)?;
+    /// assert_eq!(vec.len(), 4);
+    /// # Ok::<(), ZeroVecError>(())
+    /// ```
     pub fn len(&self) -> usize {
-        self.as_borrowed().len()
+        self.as_components().len()
```

Comment:
Thought: I was thinking that we would write custom code for each of these functions.  To get the length, we just read the first 4 bytes.  In theory, the compiler should figure that out, iff everything is inlined and DCE works fully.  But I don't know if I want to rely on that.

Reply:
I kinda like that all the implementation code is in the `components` module: Perhaps as a followup we can add helper methods there that do partial parsing?

Reply:
something like `components::len_of_buffer` and `components::indices_of_buffer(len)`

Reply:
Yes, that would work for me.